Togetherly by Helene Westrin

HeleneWestrin opened the pull request:
**API URL:** https://togetherly.up.railway.app/api
**Frontend URL:** https://togetherly-chi.vercel.app

HeleneWestrin committed:
Created basic folder structure for the backend. Added User model and authentication using JWT.

HeleneWestrin committed:
Created folder structure and started setting up user creation and login, and a component for showingrestriced data

HeleneWestrin committed:
Refactored how the auth works and added React Router and Zustand. Created function to log out. Added Tailwind.

HeleneWestrin committed:
Enhance backend and frontend functionality:
- Added wedding management features including models, services, and routes for creating and managing weddings.
- Implemented user roles and permissions for accessing wedding data.
- Introduced error handling middleware and improved authentication flow.
- Updated frontend with new components for wedding listing and details, integrated with React Query for data fetching.
- Enhanced UI with Tailwind CSS for better styling and user experience.

HeleneWestrin committed:
Implement request validation and enhance authentication middleware

- Added `validateRequest` middleware for schema validation using Zod, improving request handling in user and wedding routes.
- Enhanced `authenticateUser` middleware to include userId in the request object and improved error handling for JWT authentication.
- Introduced new Zod schemas for user and wedding data validation.
- Updated user and wedding routes to utilize the new validation middleware.
- Added new types for API responses and improved response handling in utility functions.
- Updated frontend components for better user experience and integrated new SVG assets for branding.

HeleneWestrin committed:
Enhance user management and authentication features

- Improved user creation process with detailed validation for email and password requirements.
- Added comprehensive documentation for user-related functions in the controller and service layers.
- Enhanced error handling in middleware to provide more informative responses.
- Updated user model to include profile information and guest details for weddings.
- Refined authentication middleware to ensure proper user role checks and JWT handling.
- Integrated new validation schemas for user requests using Zod.
- Updated frontend components to utilize the latest API endpoints and improve user experience with better loading states and error handling.

HeleneWestrin committed:
Refactor wedding service and enhance WeddingDetails component

- Updated the WeddingService to include only necessary fields in the query for weddings, excluding budget and adding guests.
- Enhanced the WeddingDetails component by adding a logout button, improving the layout for better user experience, and integrating logout functionality using Zustand for state management.

HeleneWestrin committed:
Refactor authentication and user management features

- Removed Babel configuration as it is no longer needed.
- Updated package.json to remove Babel dependencies and added google-auth-library for Google authentication.
- Enhanced environment configuration to include Google client ID and secret.
- Refactored user model to support social login with Google, including optional fields for social provider and ID.
- Removed unused getSecrets function from user controller and updated routes to include Google token authentication.
- Implemented Google login functionality in frontend components, allowing users to sign in with their Google accounts.
- Improved error handling and user feedback in login and account creation forms.
- Updated frontend to use new Google OAuth provider and added necessary types for social login responses.

HeleneWestrin committed:
Update .gitignore to exclude certificate files

HeleneWestrin committed:
Update .gitignore to include frontend certificate files and package-lock.json

HeleneWestrin committed:
Remove sensitive certificate files from the frontend and update .gitignore accordingly

HeleneWestrin committed:
Enhance wedding management and task functionality

- Introduced task management features, including a new task router and controller for updating task status.
- Updated wedding model to include task references and budget calculations.
- Refactored wedding service to handle task updates and budget progress calculations.
- Enhanced API routes to include task-related endpoints under the /api path.
- Improved frontend components for displaying budget categories and tasks, integrating task completion functionality.
- Updated user and wedding routes to ensure proper access control for task management.
- Refined error handling and user feedback in task-related operations.

HeleneWestrin committed:
Refactor wedding access control in middleware and service

- Enhanced weddingAuth middleware to differentiate access permissions for task operations, allowing only couples to perform task-related actions.
- Updated wedding service to include admin role checks, permitting admins to update tasks alongside couples.
- Improved error handling for user and wedding retrieval processes, ensuring clearer feedback for unauthorized access attempts.

HeleneWestrin committed:
Enhance wedding management UI and functionality

- Added 'lucide-react' dependency for icon support in the frontend.
- Updated seedDatabase function to include new task templates for 'Ceremony' and 'Entertainment' categories.
- Enhanced Button component to support 'ghost' variant and 'size' prop for better styling options.
- Improved Typography styles for better text visibility and consistency.
- Refined BudgetCategory and BudgetOverview components to enhance task display and progress indicators.
- Updated WeddingList and WeddingDetails components for improved text styling and user experience.

HeleneWestrin committed:
Implement Badge and ProgressBar components for enhanced UI feedback

- Added Badge component for displaying task progress with color-coded indicators.
- Introduced ProgressBar component for visual representation of progress in budget categories.
- Updated BudgetCategory and BudgetOverview components to utilize new Badge and ProgressBar components for improved user experience.
- Enhanced tailwind.config.js with new transition timing functions for smoother animations.

HeleneWestrin committed:
- Added a new gradient background option for mobile views in tailwind.config.js.
- Updated CSS classes to replace "bg-gradient-landscape" with "bg-gradient-full" for consistent background styling across multiple pages.
- Refactored LoginForm and FormInput components to include labels for better accessibility and user experience.
- Introduced a new TaskItem component to streamline task display within BudgetCategory, enhancing task management functionality.
- Improved transition effects in ProgressBar for smoother animations.
- Adjusted padding in WeddingDetails component for better layout consistency.

HeleneWestrin committed:
- Increased spacing in the budget section for improved visual clarity.

HeleneWestrin committed:
- Added create, update, and delete task functionalities in the WeddingController.
- Enhanced weddingAuth middleware to restrict task operations to couples and admins.
- Introduced new API routes for task management, allowing task creation and updates.
- Updated frontend components to support task creation and editing, including a new TaskForm and SidePanel for better user interaction.
- Refactored BudgetCategory and TaskItem components to integrate task management features seamlessly.
- Improved error handling and user feedback for task-related operations.
- Updated taskService to handle API interactions for task management.

HeleneWestrin committed:
- Added support for over-budget indication in BudgetOverview, displaying "Over budget" when applicable.
- Updated ProgressBar to handle progress exceeding 100%, introducing a new fill color for overflow.
- Added a new pink color variant (700) in tailwind.config.js for improved styling options.
- Cleaned up console logs in TaskForm for better code clarity.

HeleneWestrin committed:
- Added `updateBudget` method in `WeddingService` to allow budget updates by authorized users.
- Introduced `updateBudget` route in `wedding.routes.ts` for handling budget update requests.
- Enhanced `BudgetOverview` component to include a side panel for editing the total budget with validation.
- Updated `BudgetCategory` component to fetch tasks by category and display progress.
- Refactored `TaskForm` and `TaskItem` components to support currency input and improved task management.
- Added `GuestList` component for displaying wedding guests.
- Introduced `useIsMobile` hook for responsive design handling.
- Updated Tailwind CSS configuration for new color variants and improved styling.
- Enhanced error handling and user feedback across budget and task management functionalities.

HeleneWestrin committed:
- Introduced `DesktopNavigation` and `MobileNavigation` components for improved navigation experience on wedding pages.
- Added `WeddingNavigation` component to conditionally render mobile or desktop navigation based on screen size.
- Implemented `WeddingGuests` page to manage guest lists, including adding new guests with email validation.
- Updated `BudgetCategory` component to utilize new types and improve task handling.
- Refactored `GuestList` and `TaskForm` components for better type safety and user experience.
- Enhanced routing in `AppRouter` to include new wedding-related routes and navigation logic.
- Improved UI state management in `useUIStore` to handle guest addition panels.
- Updated type definitions in `wedding.ts` for better clarity and structure.
- Enhanced utility functions for wedding calculations to streamline task and budget management.

HeleneWestrin committed:
- Added `createWeddingFromOnboarding` method in `WeddingService` to facilitate wedding creation from onboarding data.
- Introduced new route in `wedding.routes.ts` for onboarding wedding creation.
- Updated `WeddingController` to handle onboarding requests and respond with success messages.
- Modified `UserService` to check for existing users during account creation, throwing a validation error if the email is already registered.
- Enhanced `CreateAccountForm` to handle new user creation and navigate based on wedding data.
- Implemented `Onboarding` page for collecting couple and wedding information.
- Updated navigation logic to redirect new users to the onboarding process.
- Improved UI components for better accessibility and user experience, including form labels and typography adjustments.
- Added new font files and updated Tailwind CSS configuration for improved styling options.

HeleneWestrin committed:
- Modified DesktopNavigation and MobileNavigation to include new icons and updated routes for user management.
- Enhanced styling in DesktopNavigation for better layout and spacing.
- Added a gradient background to the Home component for improved visual appeal.
- Updated Tailwind CSS configuration with new color variants and scrollbar styles for a more polished UI.
- Refactored BudgetCategory to improve accessibility with dynamic IDs and aria attributes for better screen reader support.

HeleneWestrin committed:
- Added OnboardingController and routes to manage onboarding progress
- Created OnboardingProgress model to track user's onboarding steps
- Enhanced Onboarding page with multi-step form, localStorage persistence, and backend synchronization
- Updated TypeScript configuration to include custom type roots
- Introduced RadioButtonToggle component for improved form interactions
- Refactored wedding and user models to use more descriptive type names
- Improved form validation and user experience in onboarding process

HeleneWestrin committed:
- Implemented `completeOnboarding` method in UserService and UserController
- Added new route `/complete-onboarding` to update user's isNewUser status
- Updated frontend onboarding flow to mark user as no longer new
- Enhanced authentication store with `updateUser` method
- Improved navigation logic for new users
- Refactored CreateAccount and Login pages with improved UI and navigation links

HeleneWestrin committed:
Add WeddingSchedule and WeddingUsers pages, update routing and preload background images

HeleneWestrin committed:
Enhance guest management with comprehensive guest data and validation

HeleneWestrin committed:
- Updated WeddingHeader component to be more flexible and reusable
- Introduced WeddingSummary component to separate wedding details display
- Enhanced guest and budget pages with new header and layout improvements
- Improved form data validation and cleaning in guest and user management
- Added support for optional email and trimming of input fields

HeleneWestrin committed:
- Adjusted navigation, typography, and button components for 2xl screen responsiveness
- Simplified budget update logic in WeddingBudget page
- Enhanced WeddingHeader and routing for better mobile/desktop experience
- Improved task creation with optional due date handling

HeleneWestrin committed:
- Enhanced GuestList with empty state handling and Typography component
- Improved Onboarding page with wedding date validation and input restrictions
- Updated WeddingBudget page with more robust rendering and error handling
- Refined guest and budget page layouts with responsive design tweaks

HeleneWestrin committed:
- Introduced skeleton loaders for budget categories and tasks
- Added Radix UI tooltips to task action buttons
- Implemented ScrollToTop component for page navigation
- Updated Button component to use forwardRef

HeleneWestrin committed:
- Added backend routes and services for updating and deleting guests
- Implemented guest role selection with predefined role options
- Created advanced GuestList component with sorting, multi-select, and bulk actions
- Updated user and wedding models to support more detailed guest information
- Enhanced AddGuestForm with improved validation and role selection
- Added tooltips and icon buttons for better user interaction

HeleneWestrin committed:
- Implemented backend routes and services for inviting users to weddings
- Created InviteUserForm and UserList components for comprehensive user management
- Added support for inviting users as couple members or guests with specific wedding roles
- Enhanced user registration flow with improved status tracking
- Updated frontend and backend to handle user invitation and registration states

HeleneWestrin committed:
- Added backend routes and services for updating and deleting wedding users
- Created EditUserForm component for user profile editing
- Enhanced UserList with edit and delete actions for wedding users
- Updated user and wedding models to support more flexible user roles
- Implemented frontend mutations for user update and deletion
- Added support for existing guest selection in invite user flow

HeleneWestrin committed:
Improved SidePanel and GuestList UI with enhanced responsiveness and guest details

HeleneWestrin committed:
Configure CORS with frontend URL and credentials

HeleneWestrin committed:
Update Vite config for production and development environments

HeleneWestrin committed:
Add Vercel Speed Insights and optimize React Query configuration

HeleneWestrin committed:
Add logging for CORS allowed origin configuration

HeleneWestrin committed:
Prepared ENV variables for production

HeleneWestrin committed:
Added vercel.json

HeleneWestrin committed:
Vercel update

HeleneWestrin committed:
Add database performance optimizations with MongoDB indexing

HeleneWestrin committed:
- Add offlineStorage service for handling offline task updates
- Implement debounced task update with offline fallback
- Update TaskItem to use new FormCheckbox component

HeleneWestrin committed:
Onbording tweaks

HeleneWestrin committed:
Improve error handling and button state in CreateAccountForm and Button components

HeleneWestrin committed:
Refine UI components and responsive design across multiple pages

HeleneWestrin committed:
FormInput styling

HeleneWestrin committed:
Enhance date input with clear button and input constraints

HeleneWestrin committed:
Skeleton loader fix

HeleneWestrin committed:
Add loading and success states with SVG spinners across forms

HeleneWestrin committed:
Add BudgetOverviewSkeleton for loading state in wedding budget page

HeleneWestrin committed:
Sidepanel

HeleneWestrin committed:
Height

HeleneWestrin committed:
Update min-height to use svh and adjust Home page button styles

HeleneWestrin committed:
Refine date input styling and form button states

HeleneWestrin committed:
Add Hotjar tracking and new WeddingMore page with routing

HeleneWestrin committed:
Copy

HeleneWestrin committed:
Config

HeleneWestrin committed:
Styling

HeleneWestrin committed:
tel

HeleneWestrin committed:
Added comments

HeleneWestrin committed:
Major refactor to wedding and user models to support more flexible access and role management.

HeleneWestrin committed:
Enhance wedding and user models with more detailed role and access management

HeleneWestrin committed:
Faster debounce delay

HeleneWestrin committed:
Task optimization

HeleneWestrin committed:
Budget category

HeleneWestrin committed:
Hopefully better performance in prod

HeleneWestrin committed:
Invite users fix

HeleneWestrin committed:
Small fix

HeleneWestrin committed:
Improve onboarding flow and login handling with new user detection

HeleneWestrin committed:
Add loading skeletons and improved loading states for wedding pages

HeleneWestrin committed:
Refactor backend services with improved error handling and lean query optimization

HeleneWestrin committed:
Cleanup

HeleneWestrin committed:
README

HeleneWestrin committed:
Refactor error handling and controller/service structures

- Replaced class-based error handling with functional error creation methods
- Converted controllers and services from class-based to functional exports
- Simplified error creation with more flexible and lightweight approach
- Standardized error handling across middleware and services
- Removed unnecessary class-based abstractions in favor of more modular functions

HeleneWestrin committed:
Refactor authentication and user management across backend and frontend

- Update authentication middleware to use more type-safe and explicit user attachment
- Modify Express types to support user property on request
- Normalize user data in auth store with more flexible structure
- Improve login hook and user type definitions
- Update wedding users page with more descriptive text
- Standardize user ID extraction in controllers using req.user.id

HeleneWestrin committed:
Styling

JennieDalgren commented on .cursor/rules/project-rules.mdc:
```diff
@@ -0,0 +1,68 @@
+---
```
Nice that you have used cursor - that's the modern way of coding, speeding up the process.

JennieDalgren commented on backend/src/config/env.ts:
```diff
@@ -0,0 +1,35 @@
+// Load environment variables from .env file
+dotenv.config();
+
+const envSchema = z.object({
+  BACKEND_URL: z.string().min(1, "BACKEND_URL is required"),
+  FRONTEND_URL: z.string().min(1, "FRONTEND_URL is required"),
+  GOOGLE_CLIENT_ID: z.string().min(1, "GOOGLE_CLIENT_ID is required"),
+  GOOGLE_CLIENT_SECRET: z.string().min(1, "GOOGLE_CLIENT_SECRET is required"),
+  JWT_SECRET: z.string().min(1, "JWT_SECRET is required"),
+  MONGO_URI: z.string().min(1, "MONGO_URI is required"),
+  NODE_ENV: z
+    .enum(["development", "production", "test"])
+    .default("development"),
+  PORT: z.string().default("8080"),
+  SEED_DATABASE: z.string().default("false"),
+});
```
was there a specific reason to why you chose to use zod?

HeleneWestrin replied:
I had used it in a previous project, and also it seemed popular 😅

JennieDalgren commented on backend/src/controllers/wedding.controller.ts:
```diff
@@ -0,0 +1,399 @@
+    const wedding = await WeddingService.createWedding(weddingInfo);
+
+    // Add wedding role for the creating user with defaults for couple users
+    await User.findByIdAndUpdate(userId, {
+      $push: {
+        weddings: {
+          weddingId: wedding._id,
+          accessLevel: "couple",
+          coupleDetails: { role: "Couple" },
+          guestDetails: {
+            rsvpStatus: RSVPStatus.CONFIRMED,
+            partyRole: WeddingPartyRoles.COUPLE,
+          },
+        },
+      },
+    });
+
```
⭐

JennieDalgren reviewed (commented):
Really good job with this project! You should be proud. You have met the requirements and beyond, adding tailwind, typescript etc.

Your code is well structured, you have used commenting to explain your code which is very helpful. However your code it self is easy to follow and understand as well. 
It looks like it became a fairly backend heavy project, and you managed in a very good way.

The product it self has a lot of nice features, the design is sleek and it really feels like a "real" product. I hope you will continue on with this project (and I hope I get to use it one day)

Keep up the great job, looking forward to follow your journey, from a distance, going forward ⭐ 

Bug:
<img width="831" alt="image" src="https://github.com/user-attachments/assets/b14552cf-a663-454f-b9d6-1ed9c93c6f47" />

When editing my name, the space between first and last name got removed

A tiny thing, suggestion to be able to change roles if there are two wives, or husbands.

JennieDalgren commented:
Another issue I just saw:
I had a wedding admin named natalie, bridesmaid. But now, it shows this: 

<img width="773" alt="image" src="https://github.com/user-attachments/assets/bd6e300e-f025-497c-9824-310308491f02" />

HeleneWestrin committed:
Implement comprehensive onboarding refactoring and improvements

- Add new onboarding hook, store, and service for better state management
- Create dedicated types for onboarding data
- Refactor onboarding page to use new hook and centralized state management
- Add API endpoint for deleting onboarding progress
- Enhance wedding invitation and user management logic
- Improve navigation and user onboarding flow

HeleneWestrin commented:
> Really good job with this project! You should be proud. You have met the requirements and beyond, adding tailwind, typescript etc.

Thank you @JennieDalgren! ☺️ It was a bit challenging with both TypeScript and the backend logic, but Cursor was a good sounding board there.

Perfect that you found some bugs also! I hopefully fixed the most severe one, where a random person showed up as Wedding Admin 😅

Thank you for your ideas! And also this whole bootcamp. It has been a blast 🤩
